### ISSUE TITLE
Generate Servicer?

### ISSUE THREAD

**Issue by jyggen** [danielgtaylor/python-betterproto#19]
Currently it seems like this package is intended for gRPC _clients_ written in Python, meaning that people who also write their gRPC _servers_ in Python will have to generate their Python bindings the old fashioned way as well as better protos for their clients. Is it on the roadmap to generate a servicer as well? Would be nice to be able to use this package on both sides.

**Comment by zidik** [comment]
I was just about to write the same question. I was looking for the ````GreeterBase```` class to extend like in this server implementation.
https://grpclib.readthedocs.io/en/latest/#server

**Comment by nat-n** [comment]
I've had success using betterproto with grpcio. It required a bit of boiler plate, and manually coding the package name which betterproto doesn't include in the generated data classes,  but that can be hidden away with a bit of meta-programming.

**Comment by githorse** [comment]
@nat-n Could you elaborate a bit on what you mean by "using betterproto with grpcio"? If I'm building a Python servicer, there's no getting around the fact that I have to compile all my protos twice (once with standard tools for server side, and once with betterproto for client side), right? Or is there some way I can return betterproto-compiled messages from my standard servicer? E.g.

```
import my_standard_grpc_module as standard
import my_betterproto_module as better

class EchoServicer(standard.EchoServicer):
    def Echo(self, request: better.EchoRequest, context) -> better.EchoResponse:
        return better.EchoResponse(message="boo")
```

(The above approach fails with `"Failed to serialize response!"` error.)

**Comment by nat-n** [comment]
@githorse IMHO the idea of generating case classes for servicers doesn't really translate well into python where the type system doesn't naturally play the same role as in Java & co.

The point is that since python lends itself better to meta-programming and it all works with duck typing there's nothing to stop you from writing your own generic glue code to shove betterproto types into the grpcio service API (as I did).

Of course it would be useful if betterproto provided this glue code, and I'd be happy to clean up and contribute a PR for my own implementation if I had confidence the PR would get accepted.

I've created [a gist demonstrating the basics](https://gist.github.com/nat-n/e90097ebfb861cbb25e20b68bec0e39c) in a way that you can probably copy easily enough. Let me know of it works for you.

**Comment by w4rum** [comment]
I really like the gist above. My main problem with it is that - if you want to use it a quick work-around at the moment - it provides a non-async interface. I've created a [modified, fully async version based on grpclib](https://gist.github.com/w4rum/4e20ec18b9065b1b6780e2f92ac4b6f0). I don't know how this could be integrated into auto-generation. I just hope that this gist might provide a good stepping stone.

**Comment by skywalkerisnull** [comment]
@w4rum I am currently looking through the PR that you did and am just trying to figure out how to specific things like the secure port, insecure port, and tell the grpc server to start? It is not immediately jumping out at me as to how to configure the server. Any pointers that you can give to setup the server?

**Comment by skywalkerisnull** [comment]
I figured it out after digging further into grpclib and this document: https://grpclib.readthedocs.io/en/latest/

**Comment by w4rum** [comment]
@skywalkerisnull 
Sorry for the late reponse. Good thing you've figured it out. There is also an [old version of a test](https://github.com/danielgtaylor/python-betterproto/blob/a54d0ce8fa0f8d957cd0ad4c6a60825c46a14aac/tests/inputs/example_service/test_example_service.py#L54) (line 54-63) that started the server without TLS. I don't know how to plug TLS straight into grpclib, as I usually just put all of my applications behind nginx and let nginx take care of TLS.

Regarding the old test, please note that the structure of the generated sources has changed and that test will no longer work. I'm just referencing the old version because that still had a pretty standard grpclib server setup.

**Comment by w4rum** [comment]
@nat-n I suppose this issue can be closed now that #170 is merged. Do you want me to write some documentation into the readme and make another PR for that?

**Comment by nat-n** [comment]
Thanks @w4rum. Yes having some help with the documentation would be great. Currently documentations is kind of duplicated over both the README, and some sphinx docs in `/docs`, ideally they should be consolidated, but otherwise I guess it would b best to maintain both for now.
